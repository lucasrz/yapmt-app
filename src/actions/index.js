import {
  SET_PROJECTS,
  SET_PROJECT,
  GET_PROJECTS_SAGA,
  CREATE_PROJECT_SAGA,
  DELETE_PROJECT_SAGA,
  ADD_TASK_SAGA,
  SET_TASK,
  UPDATE_TASK_SAGA,
  UPDATE_PROJECT_STATUS,
  SET_TASKS,
} from '../constants';

export function setProjects(projects) {
  return {
    type: SET_PROJECTS,
    projects
  };
}

export function setProject(project) {
  return {
    type: SET_PROJECT,
    project
  };
}

export function updateProjectStatus(project) {
  return {
    type: UPDATE_PROJECT_STATUS,
    status: project.status
  };
}

export function setTask(task) {
  return {
    type: SET_TASK,
    task
  };
}

export function setTasks(state) {
  return {
    type: SET_TASKS,
    tasks: state.tasks
  };
}


export function getProjectsSaga() {
  return {
    type: GET_PROJECTS_SAGA,
  };
}

export function createProjectSaga(project) {
  return {
    type: CREATE_PROJECT_SAGA,
    project
  };
}

export function deleteProjectsSaga(project) {
  return {
    type: DELETE_PROJECT_SAGA,
    project
  };
}

export function addTaskSaga(task) {
  return {
    type: ADD_TASK_SAGA,
    task
  };
}

export function updateTaskSaga(task) {
  return {
    type: UPDATE_TASK_SAGA,
    task
  };
}
