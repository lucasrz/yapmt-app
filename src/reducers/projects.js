import { SET_PROJECT, SET_PROJECTS, REQUEST_ERROR, REQUEST_SUCCESS, CREATE_PROJECT, UPDATE_PROJECT_STATUS } from '../constants';

const initialState = { projects: [] };

export default function setBrowserInfo(state = initialState, action) {
  switch (action.type) {
    case SET_PROJECTS:
      return {
        ...state,
        projects: action.projects
      };
    case SET_PROJECT:
      return {
        ...state,
        project: action.data ? action.data : action.project
      };
    case REQUEST_ERROR:
      return {
        ...state,
        error: action
      }
    case REQUEST_SUCCESS:
      return {
        ...state,
        requestStatus: action
      }
    case CREATE_PROJECT:
      return {
        ...state,
        newProject: action 
      }
    case UPDATE_PROJECT_STATUS:
      return {
        ...state,
        status: action.status
      }
    default:
      return state;
  }
}

