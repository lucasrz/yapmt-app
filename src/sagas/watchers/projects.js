import { put, takeLatest, call } from 'redux-saga/effects';
import { setProjects } from '../../actions';
import { createProject, getProjects, deleteProject } from '../../services/project-service';
import { GET_PROJECTS_SAGA, CREATE_PROJECT_SAGA, DELETE_PROJECT_SAGA, REQUEST_ERROR, SET_PROJECT  } from '../../constants';

function* workerGetProjects() {
  const projects = yield call(getProjects);
  yield put(setProjects(projects));
}

function* workerCreateProject(action) {
  try { 
    const data = yield call(createProject, action);
    yield call(workerGetProjects, action);
    yield put({ type: SET_PROJECT, data });
  } catch (error) {
    yield put({ type: REQUEST_ERROR, error })
  }
}

function* workerDeleteProject(action) {
  try { 
    yield call(deleteProject, action);
    yield put({type: SET_PROJECT});
    yield call(workerGetProjects, action);
  } catch (error) {
    yield put({ type: REQUEST_ERROR, error })
  }
}

function* watchProjectsSaga() {
  yield takeLatest(GET_PROJECTS_SAGA, workerGetProjects);
  yield takeLatest(CREATE_PROJECT_SAGA, workerCreateProject);
  yield takeLatest(DELETE_PROJECT_SAGA, workerDeleteProject);
}

export {
  watchProjectsSaga,
  workerDeleteProject,
  workerCreateProject,
  workerGetProjects,
}