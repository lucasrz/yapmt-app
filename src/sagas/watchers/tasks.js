import { put, takeLatest, call } from 'redux-saga/effects';
import { setTask } from '../../actions';
import { addTask, updateTask } from '../../services/task-service';
import { ADD_TASK_SAGA, UPDATE_TASK_SAGA, REQUEST_ERROR, GET_PROJECTS_SAGA  } from '../../constants';

function* workerAddTask(action) {
  try { 
    let task = yield call(addTask, action.task);
    task = task.res;
    yield put(setTask(task));
    yield put({ type: GET_PROJECTS_SAGA })
  } catch (error) {
    yield put({ type: REQUEST_ERROR, error })
  }
}

function* workerUpdateTask(action) {
  yield call(updateTask, action.task);
}
 
function* watchTasksSaga() {
  yield takeLatest(ADD_TASK_SAGA, workerAddTask);
  yield takeLatest(UPDATE_TASK_SAGA, workerUpdateTask);
}

export {
  watchTasksSaga,
  workerUpdateTask,
  workerAddTask,
}
