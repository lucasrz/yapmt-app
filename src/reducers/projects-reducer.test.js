import reducer from './index';
import { setProjects, setProject } from '../actions';
import { workerGetProjects } from '../sagas/watchers/projects';
import { workerCreateProject } from '../sagas/watchers/projects';

describe('Reducers', () => {
  const project = {
    name: 'testing',
    tasks: [{}]
  }

  it('Set Projects', () => {
    const initialState = {};
    const next = reducer(initialState, setProjects([project, project, project]))
    expect(next).toMatchSnapshot()
  });

  it('Set Project', () => {
    const initialState = {};
    const next = reducer(initialState, setProject(project))
    expect(next).toMatchSnapshot()
  });

  it('Request Success', () => {
    const next = workerGetProjects();
    expect(next).toMatchSnapshot();
  });

  it('Create Project', () => {
    const next = workerCreateProject(project);
    expect(next).toMatchSnapshot();
  });
});