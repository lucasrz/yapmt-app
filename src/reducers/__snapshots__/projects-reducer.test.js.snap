// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reducers Create Project 1`] = `Object {}`;

exports[`Reducers Request Success 1`] = `Object {}`;

exports[`Reducers Set Project 1`] = `
Object {
  "projectsReducer": Object {
    "project": Object {
      "project": Object {
        "name": "testing",
        "tasks": Array [
          Object {},
        ],
      },
      "type": "SET_PROJECT",
    },
    "projects": Array [],
  },
  "tasksReducer": Object {
    "task": Object {},
  },
}
`;

exports[`Reducers Set Projects 1`] = `
Object {
  "projectsReducer": Object {
    "projects": Array [
      Object {
        "name": "testing",
        "tasks": Array [
          Object {},
        ],
      },
      Object {
        "name": "testing",
        "tasks": Array [
          Object {},
        ],
      },
      Object {
        "name": "testing",
        "tasks": Array [
          Object {},
        ],
      },
    ],
  },
  "tasksReducer": Object {
    "task": Object {},
  },
}
`;
